kalite/topic_tools/__init__.py

```diff
@@ -45,24 +42,38 @@
 
 
 # Globals that can be filled
-TOPICS          = None
+TOPICS          = {}
 CACHE_VARS.append("TOPICS")
-def get_topic_tree(force=False, props=None):
-    global TOPICS, TOPICS_FILEPATH
-    if TOPICS is None or force:
-        TOPICS = softload_json(TOPICS_FILEPATH, logger=logging.debug, raises=False)
-        validate_ancestor_ids(TOPICS)  # make sure ancestor_ids are set properly
-
-        # Limit the memory footprint by unloading particular values
-        if props:
-            node_cache = get_node_cache()
-            for kind, list_by_kind in node_cache.iteritems():
-                for node_list in list_by_kind.values():
-                    for node in node_list:
-                        for att in node.keys():
-                            if att not in props:
-                                del node[att]
-    return TOPICS
+def get_topic_tree(force=False, annotate=False, channel=settings.CHANNEL):
+    global TOPICS, TOPICS_FILEPATHS
+    if TOPICS.get(channel) is None or force:
+        TOPICS[channel] = softload_json(TOPICS_FILEPATHS.get(channel), logger=logging.debug, raises=False)
+        validate_ancestor_ids(TOPICS[channel])  # make sure ancestor_ids are set properly
+
+    if annotate:
+        if settings.HARD_CONTENT_CACHE and not force:
+            topics = softload_json(TOPICS_FILEPATHS.get(channel) + ".cache", logger=logging.debug, raises=False)
+            if topics:
+                TOPICS[channel] = topics
+                return TOPICS[channel]
+
+        content_cache = get_content_cache()
+        def recurse_nodes(node):
+            # By default this is very charitable, assuming if something has not been annotated it is available
+            if content_cache.get(node.get("id"), {}).get("languages", [""]):
+                node["available"] = True
+            # Do the recursion
+            for child in node.get("children", []):
+                recurse_nodes(child)
+        recurse_nodes(TOPICS[channel])
+        if settings.HARD_CONTENT_CACHE:
```

Comment:
It gets invalidated when force and annotate are both set to true. This happens on video download, will have to double check deletion.